### PR TITLE
feat: remove product config & clean up

### PR DESCRIPTION
**Pull request by maltesander**
## Description

- remove product config
- move to config map builder step

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

### Author

- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
- [ ] Links to generated (nightly) docs added
- [ ] Release note snippet added

### Reviewer

- [x] Code contains useful comments
- [x] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)

### Acceptance

- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] Links to generated (nightly) docs added
- [ ] Release note snippet added
- [ ] Add `type/deprecation` label & add to the [deprecation schedule](https://github.com/orgs/stackabletech/projects/44/views/1)
- [ ] Add `type/experimental` label & add to the [experimental features tracker](https://github.com/orgs/stackabletech/projects/47)

### CODE REVIEW

File: rust/operator-binary/src/config/mod.rs
Comment (siegfriedweber) [inline rust/operator-binary/src/config/mod.rs]: Every operator contains similar code. Can we change `to_hadoop_xml` and the other writer functions to accept `(String, String)` instead of `(String, Option<String>)`?

File: rust/operator-binary/src/hdfs_controller.rs [hunk 1]
```diff
@@ -47,44 +38,84 @@ use stackable_operator::{
...
+/// The validated cluster: proves that config merging and validation succeeded
+/// for every role and role group before any resources are created. Placed in the
+/// controller so that subsequent steps that reference this struct only depend on
+/// the controller.
 #[derive(Clone, Debug)]
 pub struct ValidatedCluster {
+    /// The logical (and Kubernetes object) name of the cluster.
+    pub name: ClusterName,
+    /// The cluster namespace, used to build kerberos principals.
+    pub namespace: Option<String>,
```
Comment (siegfriedweber) [inline rust/operator-binary/src/hdfs_controller.rs]: Some unnecessary error handling can be avoided if `namespace` were not optional.

File: rust/operator-binary/src/controller/build/properties/hadoop_policy.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    #[test]
+    fn overrides_are_rendered_as_properties() {
+        assert_eq!(
+            build(config_overrides(&[("security.client.protocol.acl", "*")])),
+            "<?xml version=\"1.0\"?>\n<configuration>\n  \
+             <property>\n    <name>security.client.protocol.acl</name>\n    \
+             <value>*</value>\n  </property>\n</configuration>"
+        );
```
Comment (siegfriedweber) [inline rust/operator-binary/src/controller/build/properties/hadoop_policy.rs:39]: The expected XML result would be more readable with the proper format in the source code, e.g.


```suggestion
        assert_eq!(
            build(config_overrides(&[("security.client.protocol.acl", "*")])),
            concat!(
                "<?xml version=\"1.0\"?>\n",
                "<configuration>\n",
                "  <property>\n",
                "    <name>security.client.protocol.acl</name>\n",
                "    <value>*</value>\n",
                "  </property>\n",
                "</configuration>"
            )
        );
```

File: rust/operator-binary/src/controller/build/properties/mod.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        .filter_map(|(key, value)| value.map(|value| (key, value)))
+}
+
+/// Resolve user-provided [`KeyValueConfigOverrides`] into the key/value pairs to merge
+/// into a config file, dropping entries whose value is unset (`None`).
+fn resolved_overrides(
+    overrides: KeyValueConfigOverrides,
+) -> impl Iterator<Item = (String, String)> {
+    defined_entries(overrides.overrides)
+}
```
Comment (siegfriedweber) [inline rust/operator-binary/src/controller/build/properties/mod.rs]: Can we get rid of the optional values already in the validate step, so that we do not have to convert between `String` and `Option<String>` in so many places?

File: rust/operator-binary/src/controller/validate.rs [hunk 4]
```diff
@@ -62,54 +60,105 @@ pub fn validate_cluster(
...
+                    .iter()
+                    .map(|(k, v)| (k.clone(), v.clone())),
+            );
+            env_overrides.extend(
+                role_group
+                    .config
+                    .env_overrides
+                    .iter()
+                    .map(|(k, v)| (k.clone(), v.clone())),
+            );
```
Comment (siegfriedweber) [inline rust/operator-binary/src/controller/validate.rs]: **nit:** The code is sometimes overly complex:


```suggestion
            env_overrides.extend(iter: role.config.env_overrides.clone());
            env_overrides.extend(iter: role_group.config.env_overrides.clone());
```

File: rust/operator-binary/src/controller/validate.rs [hunk 5]
```diff
@@ -62,54 +60,105 @@ pub fn validate_cluster(
...
-                .merged_config(hdfs, rolegroup_name)
+                .merged_config(hdfs, &role_group_name)
                 .context(FailedToResolveConfigSnafu)?;
 
+            // Rack awareness topology labels, namenode only.
+            if hdfs_role == HdfsNodeRole::Name {
+                if let Some(rack_awareness) = hdfs.rackawareness_config() {
+                    env_overrides.insert("TOPOLOGY_LABELS".to_string(), rack_awareness);
+                }
+            }
```
Comment (siegfriedweber) [inline rust/operator-binary/src/controller/validate.rs]: This is not a user-provided env override. It should just be set in the build step and not in the validate step.

Additionally, the user is not able anymore to override it.

File: rust/operator-binary/src/hdfs_controller.rs [hunk 6]
```diff
@@ -47,44 +38,84 @@ use stackable_operator::{
...
+    /// ConfigMaps and initialization checks. Resolved once during validation.
+    pub namenode_podrefs: Vec<HdfsPodRef>,
+    /// Pod references for all journalnodes across all role groups. Resolved once
+    /// during validation.
+    pub journalnode_podrefs: Vec<HdfsPodRef>,
+}
+
+/// Cluster-wide settings resolved once during validation, so the build steps no
+/// longer need the raw `HdfsCluster` to render config. The flags are computed by
+/// the same `HdfsCluster` predicates used previously, just resolved up-front.
```
Comment (siegfriedweber) [inline rust/operator-binary/src/hdfs_controller.rs]: > previously

Referring to the previous state in comments would require the reader to know the operator's history.

File: rust/operator-binary/src/hdfs_controller.rs [hunk 7]
```diff
@@ -47,44 +38,84 @@ use stackable_operator::{
...
+    pub cluster_config: ValidatedClusterConfig,
     pub role_groups: BTreeMap<HdfsNodeRole, BTreeMap<String, ValidatedRoleGroupConfig>>,
     pub role_configs: BTreeMap<HdfsNodeRole, ValidatedRoleConfig>,
-    pub hdfs_opa_config: Option<HdfsOpaConfig>,
+    /// Pod references for all namenodes across all role groups. Needed for all
+    /// ConfigMaps and initialization checks. Resolved once during validation.
+    pub namenode_podrefs: Vec<HdfsPodRef>,
+    /// Pod references for all journalnodes across all role groups. Resolved once
+    /// during validation.
+    pub journalnode_podrefs: Vec<HdfsPodRef>,
```
Comment (siegfriedweber) [inline rust/operator-binary/src/hdfs_controller.rs]: The PodRefs are not validated. They can be calculated in the build step. The required information is already contained in `ValidatedCluster`. It is also infallible if the namespace is not optional.

File: rust/operator-binary/src/hdfs_controller.rs [hunk 8]
```diff
@@ -47,44 +38,84 @@ use stackable_operator::{
...
+/// Cluster-wide settings resolved once during validation, so the build steps no
+/// longer need the raw `HdfsCluster` to render config. The flags are computed by
+/// the same `HdfsCluster` predicates used previously, just resolved up-front.
+#[derive(Clone, Debug)]
+pub struct ValidatedClusterConfig {
+    pub dfs_replication: u8,
+    pub https_enabled: bool,
+    pub kerberos_enabled: bool,
+    pub authentication_enabled: bool,
+    pub authorization_enabled: bool,
```
Comment (siegfriedweber) [inline rust/operator-binary/src/hdfs_controller.rs]: Booleans seem to be okay in the current implementation. But usually, the configuration is also required. If using booleans here, then there will probably be code similar to:

```rust
if cluster_config.is_x_enabled {
  let x = cluster_config.get_x().expect("x must be set because we checked with is_x_enabled");
}
```

Therefore, I prefer `Option<x>` and try to avoid "boolean blindness".

File: Cargo.toml [hunk 9]
```diff
@@ -26,10 +25,11 @@ serde_json = "1.0"
 serde_yaml = "0.9"
 snafu = "0.9"
 strum = { version = "0.28", features = ["derive"] }
-tokio = { version = "1.40", features = ["full"] }
+tokio = { version = "1.52", features = ["full"] }
```
Comment (siegfriedweber) [inline Cargo.toml:28]: Should we keep the versions in sync with stackable-operator?

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/controller/validate.rs (hunk 5 above)
Comment (maltesander) [reply rust/operator-binary/src/controller/validate.rs]: yeah https://github.com/stackabletech/hdfs-operator/pull/790/commits/8a8edd0ed946bee98e8e1aedc415e4d4f707f69a

File: rust/operator-binary/src/hdfs_controller.rs (hunk 1 above)
Comment (maltesander) [reply rust/operator-binary/src/hdfs_controller.rs]: https://github.com/stackabletech/hdfs-operator/pull/790/commits/fd1fca7b82596374ee49ee75398cb0136a35eaa9

File: rust/operator-binary/src/config/mod.rs
Comment (maltesander) [reply rust/operator-binary/src/config/mod.rs]: Yeah i think that makes sense, or at least offer both variants.

File: rust/operator-binary/src/hdfs_controller.rs (hunk 6 above)
Comment (maltesander) [reply rust/operator-binary/src/hdfs_controller.rs]: Removed https://github.com/stackabletech/hdfs-operator/pull/790/commits/6bfa3fd30f07438354dc5b8514294b00d2625869

File: Cargo.toml (hunk 9 above)
Comment (maltesander) [reply Cargo.toml:28]: Yeah but i would bump op-rs, basically updated all dependencies in the PRs.

File: rust/operator-binary/src/hdfs_controller.rs (hunk 7 above)
Comment (maltesander) [reply rust/operator-binary/src/hdfs_controller.rs]: https://github.com/stackabletech/hdfs-operator/pull/790/commits/b16ef8b5e0098200d1ea62e29a74f5a7537d8204

File: rust/operator-binary/src/hdfs_controller.rs (hunk 8 above)
Comment (maltesander) [reply rust/operator-binary/src/hdfs_controller.rs]: https://github.com/stackabletech/hdfs-operator/pull/790/commits/5c309a2abe3b1ff78a546f357cf465926e3217bc

File: rust/operator-binary/src/controller/build/properties/hadoop_policy.rs (hunk 2 above)
Comment (siegfriedweber) [reply rust/operator-binary/src/controller/build/properties/hadoop_policy.rs:39]: Almost all inlined XML code would be more readable with such a change.

File: rust/operator-binary/src/controller/validate.rs [hunk 10]
```diff
@@ -62,54 +65,87 @@ pub fn validate_cluster(
...
-                rolegroup_name.clone(),
+                role_group_name,
                 ValidatedRoleGroupConfig {
+                    replicas,
                     merged_config,
-                    product_config_properties: rolegroup_config.clone(),
+                    config_overrides,
+                    env_overrides,
                 },
             );
```
Comment (siegfriedweber) [inline rust/operator-binary/src/controller/validate.rs]: This part is equivalent to https://github.com/stackabletech/opensearch-operator/blob/26.3.0/rust/operator-binary/src/controller/validate.rs#L156-L165.

In my opinion, the `collect_role_group_overrides` should be removed. It does not only collect the config overrides. Instead, I would create a function `validate_role_group_config` which returns a `ValidatedRoleGroupConfig`, equal to https://github.com/stackabletech/opensearch-operator/blob/26.3.0/rust/operator-binary/src/controller/validate.rs#L189-L256.

`ValidatedRoleGroupConfig` is equal to https://github.com/stackabletech/opensearch-operator/blob/26.3.0/rust/operator-binary/src/controller.rs#L149-L150 and https://github.com/stackabletech/opensearch-operator/blob/26.3.0/rust/operator-binary/src/framework/role_utils.rs#L34-L51. I like the approach in the OpenSearch operator because it resembles `RoleGroup` in stackable-operator.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/controller/validate.rs (hunk 10 above)
Comment (maltesander) [reply rust/operator-binary/src/controller/validate.rs]: tried here https://github.com/stackabletech/hdfs-operator/pull/790/commits/8f032c852da4811692294609caef37a316ea73df

File: rust/operator-binary/src/controller/build/properties/hadoop_policy.rs (hunk 2 above)
Comment (maltesander) [reply rust/operator-binary/src/controller/build/properties/hadoop_policy.rs:39]: https://github.com/stackabletech/hdfs-operator/pull/790/commits/833e1d487fc90b0bf90fd5dff6fbdf0178ee8317

File: rust/operator-binary/src/controller/build/properties/mod.rs (hunk 3 above)
Comment (maltesander) [reply rust/operator-binary/src/controller/build/properties/mod.rs]: https://github.com/stackabletech/hdfs-operator/pull/790/commits/5738ac8e0683f72e23aed277d0a6db8d27d1ec7b

File: rust/operator-binary/src/controller/mod.rs [hunk 11]
```diff
@@ -1,2 +1,222 @@
...
+/// (role <- role group) `configOverrides`, `envOverrides`, `cliOverrides` and
+/// `podOverrides`.
+///
+/// This is the local-`framework` [`RoleGroupConfig`](crate::framework::role_utils::RoleGroupConfig)
+/// specialised for HDFS: the validated config is the per-role [`AnyNodeConfig`],
+/// the product-specific common config is [`JavaCommonConfig`] (whose JVM-argument
+/// merge is fallible, hence the vendored framework variant), and the config
+/// overrides are [`v1alpha1::HdfsConfigOverrides`]. The `replicas` field is used
+/// to derive the per-pod [`HdfsPodRef`]s via [`ValidatedCluster::pod_refs`] and
+/// `env_overrides` is the typed [`EnvVarSet`](stackable_operator::v2::builder::pod::container::EnvVarSet).
```
Comment (siegfriedweber) [inline rust/operator-binary/src/controller/mod.rs]: `replicas` and the override fields are not part of this type definition and should be removed from the comment.

File: rust/operator-binary/src/hdfs_controller.rs [hunk 12]
```diff
@@ -567,259 +442,25 @@ pub async fn reconcile_hdfs(
...
-    merged_config: &AnyNodeConfig,
-    namenode_podrefs: &[HdfsPodRef],
+    rolegroup_config: &ValidatedRoleGroupConfig,
     service_account: &ServiceAccount,
 ) -> HdfsOperatorResult<StatefulSet> {
     tracing::info!("Setting up StatefulSet for {:?}", rolegroup_ref);
 
+    let image = &validated.image;
+    let merged_config = &rolegroup_config.config;
+    let env_overrides = Some(&rolegroup_config.env_overrides);
```
Comment (siegfriedweber) [inline rust/operator-binary/src/hdfs_controller.rs]: `Option` is not necessary here. `EnvVarSet` can be empty.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/controller/mod.rs (hunk 11 above)
Comment (maltesander) [reply rust/operator-binary/src/controller/mod.rs]: fixed https://github.com/stackabletech/hdfs-operator/pull/790/commits/e6bcd6024bf772eb6cd70f882349c1bd943995fe

File: rust/operator-binary/src/hdfs_controller.rs (hunk 12 above)
Comment (maltesander) [reply rust/operator-binary/src/hdfs_controller.rs]: removed https://github.com/stackabletech/hdfs-operator/pull/790/commits/e6bcd6024bf772eb6cd70f882349c1bd943995fe

**Comment by maltesander** [comment]
```
--- PASS: kuttl (964.59s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/orphaned-resources_hadoop-latest-3.4.2_zookeeper-latest-3.9.5_openshift-false (84.57s)
        --- PASS: kuttl/harness/cluster-operation_hadoop-latest-3.4.2_zookeeper-latest-3.9.5_openshift-false (126.36s)
        --- PASS: kuttl/harness/logging_hadoop-3.4.2_zookeeper-latest-3.9.5_openshift-false (229.68s)
        --- PASS: kuttl/harness/profiling_hadoop-3.4.2_zookeeper-latest-3.9.5_openshift-false (115.07s)
        --- PASS: kuttl/harness/smoke_hadoop-3.4.2_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-2_datanode-pvcs-default_listener-class-external-unstable_openshift-false (115.02s)
        --- PASS: kuttl/harness/topology-provider_hadoop-latest-3.4.2_zookeeper-latest-3.9.5_krb5-1.21.1_kerberos-backend-mit_openshift-false (165.15s)
        --- PASS: kuttl/harness/kerberos_hadoop-3.4.2_zookeeper-latest-3.9.5_krb5-1.21.1_opa-latest-1.16.2_kerberos-realm-CLUSTER.LOCAL_kerberos-backend-mit_openshift-false (638.59s)
PASS
```

**Review by siegfriedweber — DISMISSED** [review]
LGTM

**Comment by maltesander** [comment]
```
--- PASS: kuttl (2472.50s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/orphaned-resources_hadoop-latest-3.5.0_zookeeper-latest-3.9.5_openshift-false (130.23s)
        --- PASS: kuttl/harness/profiling_hadoop-3.5.0_zookeeper-latest-3.9.5_openshift-false (168.61s)
        --- PASS: kuttl/harness/smoke_hadoop-3.5.0_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-1_datanode-pvcs-2hdd-1ssd_listener-class-external-unstable_openshift-false (90.49s)
        --- PASS: kuttl/harness/smoke_hadoop-3.5.0_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-2_datanode-pvcs-default_listener-class-external-unstable_openshift-false (110.53s)
        --- PASS: kuttl/harness/smoke_hadoop-3.5.0_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-2_datanode-pvcs-2hdd-1ssd_listener-class-external-unstable_openshift-false (114.76s)
        --- PASS: kuttl/harness/smoke_hadoop-3.5.0_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-1_datanode-pvcs-default_listener-class-external-unstable_openshift-false (85.18s)
        --- PASS: kuttl/harness/smoke_hadoop-3.4.3_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-2_datanode-pvcs-2hdd-1ssd_listener-class-external-unstable_openshift-false (140.85s)
        --- PASS: kuttl/harness/smoke_hadoop-3.4.3_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-2_datanode-pvcs-default_listener-class-external-unstable_openshift-false (112.38s)
        --- PASS: kuttl/harness/smoke_hadoop-3.4.3_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-1_datanode-pvcs-default_listener-class-external-unstable_openshift-false (93.13s)
        --- PASS: kuttl/harness/kerberos_hadoop-3.4.3_zookeeper-latest-3.9.5_krb5-1.21.1_opa-latest-1.16.2_kerberos-realm-CLUSTER.LOCAL_kerberos-backend-mit_openshift-false (660.89s)
        --- PASS: kuttl/harness/kerberos_hadoop-3.4.3_zookeeper-latest-3.9.5_krb5-1.21.1_opa-latest-1.16.2_kerberos-realm-PROD.MYCORP_kerberos-backend-mit_openshift-false (676.03s)
        --- PASS: kuttl/harness/smoke_hadoop-3.4.3_zookeeper-3.9.5_zookeeper-latest-3.9.5_number-of-datanodes-1_datanode-pvcs-2hdd-1ssd_listener-class-external-unstable_openshift-false (83.38s)
        --- PASS: kuttl/harness/logging_hadoop-3.5.0_zookeeper-latest-3.9.5_openshift-false (301.98s)
        --- PASS: kuttl/harness/cluster-operation_hadoop-latest-3.5.0_zookeeper-latest-3.9.5_openshift-false (127.36s)
        --- PASS: kuttl/harness/topology-provider_hadoop-latest-3.5.0_zookeeper-latest-3.9.5_krb5-1.21.1_kerberos-backend-mit_openshift-false (166.58s)
        --- PASS: kuttl/harness/profiling_hadoop-3.4.3_zookeeper-latest-3.9.5_openshift-false (113.56s)
        --- PASS: kuttl/harness/kerberos_hadoop-3.5.0_zookeeper-latest-3.9.5_krb5-1.21.1_opa-latest-1.16.2_kerberos-realm-CLUSTER.LOCAL_kerberos-backend-mit_openshift-false (649.41s)
        --- PASS: kuttl/harness/logging_hadoop-3.4.3_zookeeper-latest-3.9.5_openshift-false (289.37s)
        --- PASS: kuttl/harness/kerberos_hadoop-3.5.0_zookeeper-latest-3.9.5_krb5-1.21.1_opa-latest-1.16.2_kerberos-realm-PROD.MYCORP_kerberos-backend-mit_openshift-false (625.43s)
PASS
```

**Comment by siegfriedweber** [comment]
This pull request has reached a size where it is no longer reviewable. Instead of reviewing the code changes, I just performed a black-box test. I tried to identify the features/fields that are not already covered by the integration tests. Since this pull request should only refactor the code, I expected the same results as in the main branch. For every feature, I tested only one "happy path" but no edge cases.

- [x] Object overrides
- [x] Custom CPU/memory
- [x] StorageClass override
- [x] PVC selectors
- [x] Multiple role groups
- [x] Custom PDB maxUnavailable
- [x] Disabled PDB
- [x] Graceful shutdown timeout
- [x] Custom affinity
- [x] configOverrides of all files

The Vector configuration deviates from the main branch, but the logging integration tests should have caught any errors.

**Review by siegfriedweber — APPROVED** [review]
